plugin/coc-quick-lint-js/src/index.ts

```diff
@@ -0,0 +1,122 @@
+import * as cp from 'child_process';
+import {
+  Diagnostic,
+  DiagnosticSeverity,
+  Document,
+  DocumentFilter,
+  ExtensionContext,
+  languages,
+  Range,
+  TextDocument,
+  Uri,
+  window,
+  workspace,
+} from 'coc.nvim';
+import fs from 'fs';
+import path from 'path';
+
+const documentSelector: DocumentFilter[] = [
+  {
+    language: 'javascript',
+    scheme: 'file',
+  },
+  {
+    language: 'javascript',
+    scheme: 'untitled',
+  },
+];
+
+interface QLError {
+  col: number;
+  lnum: number;
+  end_col: number;
+  end_lnum: number;
+  type: string;
+  nr: string;
+  vcol: number;
+  text: string;
+  filename: string;
+}
+
+export class QuickLintEngine {
+  private readonly source = 'quick-lint-js';
+  private diagnosticCollection = languages.createDiagnosticCollection(this.source);
+
+  constructor(private readonly context: ExtensionContext) {}
+
+  get enable(): boolean {
+    return workspace.getConfiguration(this.source).get('enable') as boolean;
+  }
+
+  get bin(): string {
+    const exe = process.platform === 'win32' ? 'quick-lint-js.cmd' : 'quick-lint-js';
+    const fullPath = path.join(this.context.extensionPath, 'node_modules', '.bin', exe);
+    if (fs.existsSync(fullPath)) {
+      return fullPath;
+    }
+
+    return '';
+  }
+
+  private parse(out: string): Diagnostic[] {
+    try {
+      const qflist = JSON.parse(out).qflist as QLError[];
+      if (!qflist.length) return [];
+
+      const diagnostics: Diagnostic[] = [];
+      for (const item of qflist) {
+        const range = Range.create(item.lnum - 1, item.col - 1, item.end_lnum - 1, item.end_col - 1);
+        const diag = Diagnostic.create(range, item.text);
+        diag.code = item.nr;
+        diag.source = this.source;
+        diag.severity = item.type === 'E' ? DiagnosticSeverity.Error : DiagnosticSeverity.Warning;
+        diagnostics.push(diag);
+      }
+      return diagnostics;
+    } catch (e) {
+      console.error(e);
+      return [];
+    }
+  }
+
+  public async lint(document: TextDocument) {
+    if (!workspace.match(documentSelector, document)) return;
+    this.diagnosticCollection.clear();
+
+    const filePath = Uri.parse(document.uri).fsPath;
+    const out = cp.spawnSync(this.bin, ['--output-format=vim-qflist-json', filePath], { encoding: 'utf8' }).stdout;
```

Comment:
Blocking: Invoking the CLI every keystroke is very inefficient. We should instead interact with quick-lint-js in-process or with a quick-lint-js server.